packages/python/accern_data/accern_data.py

```diff
@@ -574,74 +627,119 @@ def download_range(
-                expected_records.append(
-                    self._read_total(cur_date, valid_filters))
-                progress_bar.update(1)
-
-            total = sum(expected_records)
-            progress_bar.set_total(total=total)
-            progress_bar.set_description(desc="Downloading signals")
-
-            for ix, cur_date in enumerate(pd.date_range(start_date, end_date)):
-                print_fn(f"now processing {cur_date}")
-                print_fn(f"expected {expected_records[ix]}")
-                progress_bar.set_description(
-                    f"Downloading signals for {cur_date.strftime('%Y-%m-%d')}")
-                is_first_time = self._process_date(
-                    cur_date.strftime("%Y-%m-%d"),
+
+        valid_mode: Optional[Mode] = None
+        cur_date: Optional[pd.Timestamp] = None
+        prev_date: Optional[pd.Timestamp] = None
+        indicator_obj: Optional[ProgressIndicator] = None
+
+        def set_active_mode(
+                mode: Mode,
+                date: pd.Timestamp,
+                indicator: ProgressIndicator) -> None:
+            nonlocal valid_mode, cur_date, indicator_obj
+            valid_mode = mode
+            cur_date = date
+            indicator_obj = indicator
+
+        is_first_time = True
+        for res in self.iterate_range(
+                start_date=start_date,
+                end_date=end_date,
+                mode=mode,
+                filters=filters,
+                chunk_size=None,
+                indicator=indicator,
+                set_active_mode=set_active_mode):
+            assert valid_mode is not None
+            assert cur_date is not None
+            assert indicator_obj is not None
+
+            if cur_date != prev_date:
+                if prev_date is not None:
+                    valid_mode.finish_day(indicator_obj)
+                valid_mode.init_day(
+                    cur_date.strftime('%Y-%m-%d'),
                     output_path,
                     output_pattern,
-                    is_first_time=is_first_time,
-                    mode=valid_mode,
-                    filters=valid_filters,
-                    progress_bar=progress_bar)
-        progress_bar.close()
+                    is_first_time,
+                    indicator_obj)
+            valid_mode.add_result(res)
+            prev_date = cur_date
+            is_first_time = False
+
+        if prev_date is not None:
+            assert valid_mode is not None
+            assert indicator_obj is not None
+            valid_mode.finish_day(indicator_obj)
+
+    def iterate_range(
+            self,
+            start_date: str,
+            end_date: Optional[str] = None,
+            mode: Optional[
+                Union[Mode, ModeType, Tuple[ModeType, bool]]] = None,
+            filters: Optional[FiltersType] = None,
+            chunk_size: Optional[int] = None,
+            indicator: Optional[Indicators] = "pbar",
```

Comment:
do the same as for mode -- use a factory function -- either you pass the literal or the object itself. if you pass None you do default values